llvm/lib/Target/AIE/AIEBaseSubtarget.cpp

```diff
@@ -314,20 +315,102 @@ class RegionEndEdges : public ScheduleDAGMutation {
 /// "fixed" SUnits.
 class EmitFixedSUnits : public ScheduleDAGMutation {
 public:
+  struct RegAvailabilityTracker {
```

Comment:
I would even go a bit further and compute a "register event view". As a first conservative version:
 - Every instruction would create "read" events in its first cycle for every input operand
 - and "write" events in its last cycle (determined from `maxLatency()`) for every output operand

This would allow us to use the view for all cases:
 - Deps between a non-pipelined loop and its epilogue
 - Deps between top fixed and free instructions
 - Deps between free and bottom fixed instructions
 - etc.

What do you think? This could later become the base of timing-aware live ranges if we ever do register allocation after scheduling.

Reply:
The downside would be: it cannot be called RAT anymore. REV is still quite cool though.

Reply:
Humm, interesting. But I miss the point about the usage of `maxLatency()` here. For example, a post in load produces two outputs in different cycles and so I think it will be too pessimistic.

Reply:
Hi @gboss, I believe we should adopt a KISS approach for this REV, given our current requirements. Specifically, we need to have a clear understanding of:

- The last cycle in which a register is defined.
- The first cycle in which a register is read.

With this information, we can replace the `findEarliestRef` method by directly connecting free SUs to ExitSU. Although this may slightly increase the cost for `isAvailableNode` due to a higher number of pending SUs, it will simplify the process of comparing all free instructions against the BotFixed bundles to identify the first reference.

I propose creating this event view as a separate class, outside of the subtarget, so that we can extend it as needed in the future.

What are your thoughts?